File: lib/py/uapi/internal/validation/ValidateMockStub.py
from typing import Any, Dict, List, Union

from uapi.internal.types.UFn import UFn
from uapi.internal.types.UStruct import UStruct
from uapi.internal.types.UType import UType
from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
from uapi.internal.types.UUnion import UUnion
from uapi.internal.validation.GetTypeUnexpectedValidationFailure import get_type_unexpected_validation_failure
from uapi.internal.validation.ValidationFailure import ValidationFailure


def validate_mock_stub(given_obj: Any, select: Dict[str, Any], fn: str,
                       type_parameters: List['UTypeDeclaration'], generics: List['UTypeDeclaration'],
                       types: Dict[str, 'UType']) -> List['ValidationFailure']:
    validation_failures: List[ValidationFailure] = []

    if not isinstance(given_obj, dict):
        return get_type_unexpected_validation_failure([], given_obj, "Object")

    given_map: Dict[str, Any] = given_obj

    regex_string = "^fn\\..*$"

    keys = sorted(given_map.keys())

    matches = [k for k in keys if re.match(regex_string, k)]
    if len(matches) != 1:
        return [
            ValidationFailure([], "ObjectKeyRegexMatchCountUnexpected",
                              {"regex": regex_string, "actual": len(matches), "expected": 1, "keys": keys})
        ]

    function_name = matches[0]
    function_def: UFn = types[function_name]
    input = given_map[function_name]

    function_def_call: UUnion = function_def.call
    function_def_name: str = function_def.name
    function_def_call_cases: Dict[str, UStruct] = function_def_call.cases
    input_failures = function_def_call_cases[function_def_name].validate(
        input, select, fn, [], [])

    input_failures_with_path = []
    for f in input_failures:
        this_path = [function_name] + f.path

        input_failures_with_path.append(
            ValidationFailure(this_path, f.reason, f.data))

    input_failures_without_missing_required = [
        f for f in input_failures_with_path if f.reason != "RequiredObjectKeyMissing"
    ]

    validation_failures.extend(input_failures_without_missing_required)

    result_def_key = "->"

    if result_def_key not in given_map:
        validation_failures.append(ValidationFailure(
            [result_def_key], "RequiredObjectKeyMissing", {}))
    else:
        output = given_map[result_def_key]
        output_failures = function_def.result.validate(
            output, select, fn, [], [])

        output_failures_with_path = []
        for f in output_failures:
            this_path = [result_def_key] + f.path

            output_failures_with_path.append(
                ValidationFailure(this_path, f.reason, f.data))

        failures_without_missing_required = [
            f for f in output_failures_with_path if f.reason != "RequiredObjectKeyMissing"
        ]

        validation_failures.extend(failures_without_missing_required)

    disallowed_fields = [k for k in given_map.keys(
    ) if k not in matches and k != result_def_key]
    for disallowed_field in disallowed_fields:
        validation_failures.append(ValidationFailure(
            [disallowed_field], "ObjectKeyDisallowed", {}))

    return validation_failures

File: lib/java/src/main/java/uapi/internal/validation/ValidateMockStub.java
package uapi.internal.validation;

import static uapi.internal.validation.GetTypeUnexpectedValidationFailure.getTypeUnexpectedValidationFailure;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import uapi.internal.types.UFn;
import uapi.internal.types.UStruct;
import uapi.internal.types.UType;
import uapi.internal.types.UTypeDeclaration;
import uapi.internal.types.UUnion;

public class ValidateMockStub {
    public static List<ValidationFailure> validateMockStub(Object givenObj,
            List<UTypeDeclaration> typeParameters, Map<String, UType> types, ValidateContext ctx) {
        final var validationFailures = new ArrayList<ValidationFailure>();

        if (!(givenObj instanceof Map)) {
            return getTypeUnexpectedValidationFailure(List.of(), givenObj, "Object");
        }
        final Map<String, Object> givenMap = (Map<String, Object>) givenObj;

        final var regexString = "^fn\\..*$";

        final var keys = givenMap.keySet().stream().sorted().toList();

        final var matches = keys.stream().filter(k -> k.matches(regexString)).toList();
        if (matches.size() != 1) {
            return List.of(
                    new ValidationFailure(List.of(),
                            "ObjectKeyRegexMatchCountUnexpected",
                            Map.of("regex", regexString, "actual",
                                    matches.size(), "expected", 1, "keys", keys)));

        }

        final var functionName = matches.get(0);
        final var functionDef = (UFn) types.get(functionName);
        final var input = givenMap.get(functionName);

        final UUnion functionDefCall = functionDef.call;
        final String functionDefName = functionDef.name;
        final Map<String, UStruct> functionDefCallTags = functionDefCall.tags;
        final var inputFailures = functionDefCallTags.get(functionDefName).validate(input, List.of(), ctx);

        final var inputFailuresWithPath = new ArrayList<ValidationFailure>();
        for (final var f : inputFailures) {
            final List<Object> thisPath = new ArrayList<>(f.path);
            thisPath.add(0, functionName);

            inputFailuresWithPath.add(new ValidationFailure(thisPath, f.reason, f.data));
        }

        final var inputFailuresWithoutMissingRequired = inputFailuresWithPath.stream()
                .filter(f -> !Objects.equals(f.reason, "RequiredObjectKeyMissing")).toList();

        validationFailures.addAll(inputFailuresWithoutMissingRequired);

        final var resultDefKey = "->";

        if (!givenMap.containsKey(resultDefKey)) {
            validationFailures.add(new ValidationFailure(List.of(),
                    "RequiredObjectKeyMissing",
                    Map.of("key", resultDefKey)));
        } else {
            final var output = givenMap.get(resultDefKey);
            final var outputFailures = functionDef.result.validate(output, List.of(), ctx);

            final var outputFailuresWithPath = new ArrayList<ValidationFailure>();
            for (final var f : outputFailures) {
                final List<Object> thisPath = new ArrayList<>(f.path);
                thisPath.add(0, resultDefKey);

                outputFailuresWithPath.add(new ValidationFailure(thisPath, f.reason, f.data));
            }

            final var failuresWithoutMissingRequired = outputFailuresWithPath
                    .stream()
                    .filter(f -> !Objects.equals(f.reason, "RequiredObjectKeyMissing"))
                    .toList();

            validationFailures.addAll(failuresWithoutMissingRequired);
        }

        final var disallowedFields = givenMap.keySet().stream()
                .filter(k -> !matches.contains(k) && !Objects.equals(k, resultDefKey)).toList();
        for (final var disallowedField : disallowedFields) {
            validationFailures
                    .add(new ValidationFailure(List.of(disallowedField), "ObjectKeyDisallowed", Map.of()));
        }

        return validationFailures;
    }
}
